_posts/2018-01-09-dagger2-android-subcomponents.md

```diff
@@ -0,0 +1,205 @@
+---
+layout: post
+title:  "Using Dagger2 SubComponents to Propagate Dependancies"
+date:   2018-01-09 8:00:00
+author: abhishek
+categories: Android, Dependancy Injection
+---
+Dependancy Injection makes your code scalable and testable. `Dagger2` is almost de-facto Dependancy Injection framework in Java these days, and, Its supported and used by Google which makes it first choice of all Android Developers.
+
+In this post I am going to demonstrate use of `Subcomponents` in a typical Android app and show how Dependancy Injection is done from top to bottom. [In this app](https://github.com/abhishekBansal/android-mvp-retrofit2-dagger2-rxjava2-testing) user can see list of movies and click on any list item to see its details. Movie list is fetched from remote API, result of this API are cached in local database and this database is then used to present movie details on movie detail screen.
+
+#### Lets start!
+
+### Identify Top Level Dependancies
+These are the `Singletons` that you usually have in your app like `Retrofit APIModule`, `SharedPreferences`, `DatabaseWrapper`, `ApplicationContext` etc. It makes sense to initialise them once and then use everywhere, only problem with that is, they make unit testing Hell!! So in modern TDD world we avoid `Singletons`. `Singletons` are bad for health of testable code. With `Dagger2` we still initialise them once and then use it everywhere but, these dependancies are now injected instead of doing `Someclass.getInstance()`. In our movie listing app I can see following global dependancies
+1. ApiModule- Provides access to retrofit for networking in our presenters.
+2. DatabaseInteractor- Database handler which can be used for inserting and retrieving data from local database.
+3. Application- Android application context
+
+### Create Top Level Component and Modules
+Once we are done Identifying our `Singletons` we are ready to create a Application level `Component`. A `Component` in `Dagger2` is something which provides(or supplies) dependancies to injection target(i.e. the object which need dependancies). `Component`s are interfaces between `Module` and Injection Target. `Module`s are classes which do actual work of instantiating dependancies or creating actual objects which `Component` will supply to Injection Targets.
```

Comment:
Adding a diagram here will greatly help understanding. Its getting a bit complicated. Can we do this?